monoio/src/fs/file/mod.rs

```diff
@@ -1,543 +1,708 @@
+        op.result().await
+    }
+
+    /// Read the exact number of bytes required to fill `buf` at the specified
+    /// offset from the file.
+    ///
+    /// This function reads as many as bytes as necessary to completely fill the
+    /// specified buffer `buf`.
+    ///
+    /// # Return
+    ///
+    /// The method returns the operation result and the same buffer value passed
+    /// as an argument.
+    ///
+    /// If the method returns [`Ok(())`], then the read was successful.
+    ///
+    /// # Errors
+    ///
+    /// If this function encounters an error of the kind
+    /// [`ErrorKind::Interrupted`] then the error is ignored and the
+    /// operation will continue.
+    ///
+    /// If this function encounters an "end of file" before completely filling
+    /// the buffer, it returns an error of the kind
+    /// [`ErrorKind::UnexpectedEof`]. The buffer is returned on error.
+    ///
+    /// If this function encounters any form of I/O or other error, an error
+    /// variant will be returned. The buffer is returned on error.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use monoio::fs::File;
+    ///
+    /// #[monoio::main]
+    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    ///     let f = File::open("foo.txt").await?;
+    ///     let buffer = vec![0; 10];
+    ///
+    ///     // Read up to 10 bytes
+    ///     let (res, buffer) = f.read_exact_at(buffer, 0).await;
+    ///     res?;
+    ///
+    ///     println!("The bytes: {:?}", buffer);
+    ///
+    ///     // Close the file
+    ///     f.close().await?;
+    ///     Ok(())
+    /// }
+    /// ```
+    ///
+    /// [`ErrorKind::Interrupted`]: std::io::ErrorKind::Interrupted
+    /// [`ErrorKind::UnexpectedEof`]: std::io::ErrorKind::UnexpectedEof
+    pub async fn read_exact_at<T: IoBufMut>(
+        &self,
+        mut buf: T,
+        pos: u64,
+    ) -> crate::BufResult<(), T> {
+        let len = buf.bytes_total();
+        let mut read = 0;
+        while read < len {
+            let slice = unsafe { buf.slice_mut_unchecked(read..len) };
+            let (res, slice) = self.read_at(slice, pos + read as u64).await;
+            buf = slice.into_inner();
+            match res {
+                Ok(0) => {
+                    return (
+                        Err(io::Error::new(
+                            io::ErrorKind::UnexpectedEof,
+                            "failed to fill whole buffer",
+                        )),
+                        buf,
+                    )
+                }
+                Ok(n) => {
+                    read += n;
+                }
+                Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {}
+                Err(e) => return (Err(e), buf),
+            };
+        }
+
+        (Ok(()), buf)
+    }
+
+    async fn write<T: IoBuf>(&mut self, buf: T) -> crate::BufResult<usize, T> {
+        let op = Op::write(self.fd.clone(), buf).unwrap();

```

Comment:
The WriteOp(as well as ReadOp)'s legacy implementation is to directly call `pwrite`. This is fine for something like TcpStream(eq to send), but for file io this is a blocking action.
To make write be non-blocking, iouring/aio may help. But for legacy path, I think we can do it in an independent thread pool.
A (shared) thread pool can be attached to a runtime if it needs it. We can do blocking syscall with it. If it is not attached, we can do it inplace.

Reply:
you mean when the feature "sync" is enabled, use the thread pool to do the syscall, is that right?
If we do this, it may be necessary to change other file-related behavior like `create_dir` & `metadata` and ...